dips/dip-169.md

```diff
@@ -0,0 +1,261 @@
+        secondary_signer_addresses: Vec<AccountAddress>,
+        secondary_signers: Vec<AccountAuthenticator>,
+    },
+}
+```
+
+
+
+### Additional Checks During Transaction Validation
+
+In addition to signature checking, we also perform the following additional checks on signed multi-agent transactions during transaction validation:
+
+* The number of secondary authenticators in the signed transaction is the same as the number of secondary signer addresses that all parties have signed over.
+* There are no duplicates in the signers. In other words, sender and all the secondary signers have distinct account addresses.
+
+
+
+
+## Diem Framework Changes
+
+A new prologue called `multi_agent_script_prologue` is added to `DiemAccount.move` to validate multi-agent transactions. 
+```rust
+fun multi_agent_script_prologue<Token: store>(
+    sender: signer,
+    txn_sequence_number: u64,
+    txn_sender_public_key: vector<u8>,
+    secondary_signer_addresses: vector<address>,
+    secondary_signer_public_key_hashes: vector<vector<u8>>,
+    txn_gas_price: u64,
+    txn_max_gas_units: u64,
+    txn_expiration_time: u64,
+    chain_id: u8,
+)
+```
+ 
+Following are the additional checks it performs.
+
+
+### Account Existence Checking in Prologue
+
+Today we check in the prologue that the sender of the transaction has a `DiemAccount` resource under their address. With the ability to have multiple signers, we need to check that all of them have `DiemAccount` resources during the prologue.
+
+```rust
+let i = 0;
+while (i < num_secondary_signers) {
+    let secondary_address = *Vector::borrow(&secondary_signer_addresses, i);
+    assert(
+        exists_at(secondary_address), 
+        Errors::invalid_argument(PROLOGUE_EACCOUNT_DNE)
+    );
+    i = i + 1;
+};
+```
+
+
+
+### Authentication Key Checking in Prologue
+
+Currently in the prologue, we check that the hash of the sender’s public key is equal to the authentication key that the sender has on chain. In the new scheme, we will need to perform the same check for public keys of all the secondary signers as well. In the `multi_agent_script_prologue`, we add a loop going through the two vectors containing addresses and public key hashes to perform the aforementioned check. 
+
+
+```rust
+let i = 0;
+while (i < num_secondary_signers) {
+    let secondary_address = *Vector::borrow(&secondary_signer_addresses, i);
+    let signer_account = borrow_global<DiemAccount>(secondary_address);
+    let signer_public_key_hash = *Vector::borrow(&secondary_signer_public_key_hashes, i);
+    assert(
+        signer_public_key_hash == *&signer_account.authentication_key,
+        Errors::invalid_argument(PROLOGUE_EINVALID_ACCOUNT_AUTH_KEY),
+    );
+    i = i + 1;
+};
+```
+
+# Feature Gating
+
+Since there is no application of multi-agent transaction scheme in Diem yet, this feature is gated behind `DiemVersion = 10`. Multi-agent transactions are rejected when the Diem version is less than 10.
```

Comment:
why 10?

Reply:
I think we should omit the feature gating from the DIP. IMO, this proposal to use version 10 should not have ever been anything more than a temporary hack to let us land the PR while we were figuring out what, if any, kind of feature gating to use for this. That discussion still hasn't settled and it seems beyond the scope of the DIP regardless.

Reply:
I heard back from @sblackshear that he and @aching have agreed that multi-agent can be gated at Diem version 3. I have updated the DIP to reflect that.

Reply:
@emmazzz, I would agree with @bob-wilson that it might be hard to make this promise in this DIP about which version it would enabled in.  It might be useful to leave that detail until the "final" phase if we want to include it - although I also not sure whether it really belongs in a DIP at all.